From 4462a5b79e14a06eefb65b0cefcc4a4d709f81ba Mon Sep 17 00:00:00 2001
From: Nikolay Sivov <nsivov@codeweavers.com>
Date: Wed, 11 Feb 2026 10:51:10 +0100
Subject: [PATCH] d2d1: Use non-multiplied transform when forwarding
 transformed geometry methods.

If source geometry is itself a transformed geometry, same transform will get applied
twice. Another option would be to navigate the whole chain when transformed geometry
is created and forward methods to the geometry at the end of the chain.

Signed-off-by: Nikolay Sivov <nsivov@codeweavers.com>
(cherry picked from commit a4b45c08c0b0f7216591eec98a5b38e7ecf6dee2)
---
 dlls/d2d1/geometry.c   | 12 ++++++------
 dlls/d2d1/tests/d2d1.c |  5 ++---
 2 files changed, 8 insertions(+), 9 deletions(-)

diff --git a/dlls/d2d1/geometry.c b/dlls/d2d1/geometry.c
index 55d650ec354..1ade8f85548 100644
--- a/dlls/d2d1/geometry.c
+++ b/dlls/d2d1/geometry.c
@@ -5835,7 +5835,7 @@ static HRESULT STDMETHODCALLTYPE d2d_transformed_geometry_GetBounds(ID2D1Transfo
 
     TRACE("iface %p, transform %p, bounds %p.\n", iface, transform, bounds);
 
-    g = geometry->transform;
+    g = geometry->u.transformed.transform;
     if (transform)
         d2d_matrix_multiply(&g, transform);
 
@@ -5862,7 +5862,7 @@ static HRESULT STDMETHODCALLTYPE d2d_transformed_geometry_StrokeContainsPoint(ID
     TRACE("iface %p, point %s, stroke_width %.8e, stroke_style %p, transform %p, tolerance %.8e, contains %p.\n",
             iface, debug_d2d_point_2f(&point), stroke_width, stroke_style, transform, tolerance, contains);
 
-    g = geometry->transform;
+    g = geometry->u.transformed.transform;
     stroke_width /= g.m11;
     if (transform)
         d2d_matrix_multiply(&g, transform);
@@ -5883,7 +5883,7 @@ static HRESULT STDMETHODCALLTYPE d2d_transformed_geometry_FillContainsPoint(ID2D
     TRACE("iface %p, point %s, transform %p, tolerance %.8e, contains %p.\n",
             iface, debug_d2d_point_2f(&point), transform, tolerance, contains);
 
-    g = geometry->transform;
+    g = geometry->u.transformed.transform;
     if (transform)
         d2d_matrix_multiply(&g, transform);
 
@@ -5909,7 +5909,7 @@ static HRESULT STDMETHODCALLTYPE d2d_transformed_geometry_Simplify(ID2D1Transfor
     TRACE("iface %p, option %#x, transform %p, tolerance %.8e, sink %p.\n",
             iface, option, transform, tolerance, sink);
 
-    g = geometry->transform;
+    g = geometry->u.transformed.transform;
     if (transform)
         d2d_matrix_multiply(&g, transform);
 
@@ -5924,7 +5924,7 @@ static HRESULT STDMETHODCALLTYPE d2d_transformed_geometry_Tessellate(ID2D1Transf
 
     TRACE("iface %p, transform %p, tolerance %.8e, sink %p.\n", iface, transform, tolerance, sink);
 
-    g = geometry->transform;
+    g = geometry->u.transformed.transform;
     if (transform)
         d2d_matrix_multiply(&g, transform);
 
@@ -5957,7 +5957,7 @@ static HRESULT STDMETHODCALLTYPE d2d_transformed_geometry_ComputeArea(ID2D1Trans
 
     TRACE("iface %p, transform %p, tolerance %.8e, area %p.\n", iface, transform, tolerance, area);
 
-    g = geometry->transform;
+    g = geometry->u.transformed.transform;
     if (transform)
         d2d_matrix_multiply(&g, transform);
 
diff --git a/dlls/d2d1/tests/d2d1.c b/dlls/d2d1/tests/d2d1.c
index 71ac4dde4f2..d551f1e8efe 100644
--- a/dlls/d2d1/tests/d2d1.c
+++ b/dlls/d2d1/tests/d2d1.c
@@ -18072,12 +18072,12 @@ static void test_transformed_geometry(BOOL d3d11)
     area = 0.0f;
     hr = ID2D1TransformedGeometry_ComputeArea(geometry, NULL, 0.0f, &area);
     ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
-    ok(area > 0.0f, "Unexpected value %.8e.\n", area);
+    ok(compare_float(area, 50.0f, 0), "Unexpected value %.8e.\n", area);
 
     area = 0.0f;
     hr = ID2D1TransformedGeometry_ComputeArea(geometry2, NULL, 0.0f, &area);
     ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
-    ok(area > 0.0f, "Unexpected value %.8e.\n", area);
+    ok(compare_float(area, 50.0f, 0), "Unexpected value %.8e.\n", area);
 
     set_rect(&bounds, 1.0f, 2.0f, 3.0f, 4.0f);
     hr = ID2D1TransformedGeometry_GetBounds(geometry, NULL, &bounds);
@@ -18090,7 +18090,6 @@ static void test_transformed_geometry(BOOL d3d11)
     hr = ID2D1TransformedGeometry_GetBounds(geometry2, NULL, &bounds);
     ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
     match = compare_rect(&bounds, 720.0f, 2160.0f, 730.0f, 2170.0f, 0);
-    todo_wine
     ok(match, "Got unexpected bounds {%.8e, %.8e, %.8e, %.8e}.\n",
             bounds.left, bounds.top, bounds.right, bounds.bottom);
 
